src/lexparse/lexer.go

```diff
@@ -3,73 +3,228 @@ package lexparse
+    lexer.passWhitespace()
+  }
+
+  return tokens
+}
+
+func (lexer *Lexer) grabToken() Token {
+  lexer.newToken()
+
+  c := lexer.next()
+  if unicode.IsLetter(c) || (c == '_') {
+    lexer.grabSymbol()
+  } else if unicode.IsDigit(c) {
+    lexer.grabDecimalNumber()
+  } else {
+    lexer.grabPunctuation()
+  }
+
+  return lexer.currentToken()
+}
+
+func (lexer *Lexer) grabSymbol() {
+  c := lexer.next()
+  for unicode.IsLetter(c) || unicode.IsDigit(c) || (c == '_') {
+    lexer.advance()
+    c = lexer.next()
+  }
+
+  switch symbol := string(lexer.CurrentBlob); symbol {
+    case "true":  lexer.CurrentTType = BoolT
+    case "false": lexer.CurrentTType = BoolT
+    case "null":  lexer.CurrentTType = NullT
+    case "for":   lexer.CurrentTType = ForT
+    case "while": lexer.CurrentTType = WhileT
+    case "in":    lexer.CurrentTType = InT
+    case "type":  lexer.CurrentTType = TypeT
+    default:      lexer.CurrentTType = SymbolT
+  }
+}
+
+func (lexer *Lexer) grabDecimalNumber() {
+  for unicode.IsDigit(lexer.next()) {
+    lexer.advance()
+  }
+
+  if lexer.next() != '.' {
+    lexer.CurrentTType = IntT
+  } else {
+    lexer.CurrentTType = FloatT
+
+    lexer.advance()
+
+    for unicode.IsDigit(lexer.next()) {
+      lexer.advance()
+    }
+  }
+}
+
+func (lexer *Lexer) grabPunctuation() {
+  lexer.advance()
+
+  var blob string
+  twochars := string(append(lexer.CurrentBlob, lexer.next()))
+
+  if _, ok := punct_combos[twochars]; ok {
+    blob = twochars
+  } else {
+    blob = string(lexer.CurrentBlob)
+  }
+
+  // Need to check specific cases first because some
+  // (such as angle brackets) can be interpreted as
+  // multiple categories and may be reassigned a token
+  // type later based on context
+  switch blob {
+    case "(":  lexer.CurrentTType = LParT
+    case ")":  lexer.CurrentTType = RParT
+    case "[":  lexer.CurrentTType = LSquareBrT
+    case "]":  lexer.CurrentTType = RSquareBrT
+    case "{":  lexer.CurrentTType = LCurlyBrT
+    case "}":  lexer.CurrentTType = RCurlyBrT
+    case "<":  lexer.CurrentTType = LAngleT
+    case ">":  lexer.CurrentTType = RAngleT
+    case ".":  lexer.CurrentTType = DotT
+    case "?":  lexer.CurrentTType = QMarkT
+    case "..": lexer.CurrentTType = EllipsisT
+    case ",":  lexer.CurrentTType = CommaT
+    case ";":  lexer.CurrentTType = SemicolonT
+    case ":":  lexer.CurrentTType = ColonT
+    case "<:": lexer.CurrentTType = SubtypeT
+    default:
+  }
+
+  if lexer.CurrentTType == -1 {
+    if _, ok := binops[blob]; ok {
+      lexer.CurrentTType = BinopT
+    } else if _, ok := comparisons[blob]; ok {
+      lexer.CurrentTType = ComparisonT
+    } else if _, ok := setters[blob]; ok {
+      lexer.CurrentTType = SetterT
+    } else if _, ok := prefixes[blob]; ok {
+      lexer.CurrentTType = PrefixT
+    } else if _, ok := suffixes[blob]; ok {
+      lexer.CurrentTType = SuffixT
+    } else {
+      lexparsePanic(lexer.Line, lexer.Col-1, 1, "Invalid start to token")
```

Comment:
Not sure how much you care about error handling, but instead of immediately panicking you could set a special `ErrorT` token. That way, you can still catch syntax and possibly even semantic errors even if the code has lexical errors.

Reply:
Good idea!